src/_arraykit.c

```diff
@@ -365,6 +365,69 @@ resolve_dtype_iter(PyObject *Py_UNUSED(m), PyObject *arg)
 //------------------------------------------------------------------------------
 // general utility
 
+static PyObject *
+dtype_from_element(PyObject *Py_UNUSED(m), PyObject *arg)
+{
+    // -------------------------------------------------------------------------
+    // 1. Handle fast, exact type checks first.
+
+    // None
+    if (arg == Py_None) {
+        return (PyObject*)PyArray_DescrFromType(NPY_OBJECT);
+    }
+
+    // Float
+    if (PyFloat_CheckExact(arg)) {
+        return (PyObject*)PyArray_DescrFromType(NPY_DOUBLE);
+    }
+
+    // Integers
+    if (PyLong_CheckExact(arg)) {
+        return (PyObject*)PyArray_DescrFromType(NPY_LONG);
+    }
+
+    // Bool
+    if (PyBool_Check(arg)) {
+        return (PyObject*)PyArray_DescrFromType(NPY_BOOL);
+    }
+
+    PyObject* dtype = NULL;
+
+    // String
+    if (PyUnicode_CheckExact(arg)) {
+        PyArray_Descr* descr = PyArray_DescrFromType(NPY_UNICODE);
+        dtype = (PyObject*)PyArray_DescrFromObject(arg, descr);
+        if (!dtype) {
+            return NULL;
+        }
+        return dtype;
+    }
+
+    // Bytes
+    if (PyBytes_CheckExact(arg)) {
+        PyArray_Descr* descr = PyArray_DescrFromType(NPY_STRING);
+        dtype = (PyObject*)PyArray_DescrFromObject(arg, descr);
+        if (!dtype) {
+            return NULL;
+        }
```

Comment:
```suggestion
        if (descr == NULL) {
            return NULL;
        }
        dtype = (PyObject*)PyArray_DescrFromObject(arg, descr);
        Py_DECREF(descr);
```

Reply:
I didn't know this could raise an exception!